crates/red_knot_python_semantic/src/semantic_index/use_def/symbol_state.rs

```diff
@@ -0,0 +1,362 @@
+                // bitsets can never get out of sync.
+                unreachable!("definitions and constraints length mismatch");
+            };
+            merged.constraints.push(constraints.clone());
+        };
+
+        loop {
+            match (opt_a_def, opt_b_def) {
+                (Some(a_def), Some(b_def)) => match a_def.cmp(&b_def) {
+                    std::cmp::Ordering::Less => {
+                        // Next definition ID is only in `a`, push it to `merged` and advance `a`.
+                        push(a_def, &mut a_constraints_iter, &mut merged);
+                        opt_a_def = a_defs_iter.next();
+                    }
+                    std::cmp::Ordering::Greater => {
+                        // Next definition ID is only in `b`, push it to `merged` and advance `b`.
+                        push(b_def, &mut b_constraints_iter, &mut merged);
+                        opt_b_def = b_defs_iter.next();
+                    }
+                    std::cmp::Ordering::Equal => {
+                        // Next definition is in both; push to `merged` and intersect constraints.
+                        push(a_def, &mut a_constraints_iter, &mut merged);
+                        let Some(b_constraints) = b_constraints_iter.next() else {
+                            // SAFETY: see above.
+                            unreachable!("definitions and constraints length mismatch");
+                        };
+                        // If the same definition is visible through both paths, any constraint
+                        // that applies on only one path is irrelevant to the resulting type from
+                        // unioning the two paths, so we intersect the constraints.
+                        merged
+                            .constraints
+                            .last_mut()
+                            .unwrap()
+                            .intersect(b_constraints);
+                        opt_a_def = a_defs_iter.next();
+                        opt_b_def = b_defs_iter.next();
+                    }
+                },
+                (Some(a_def), None) => {
+                    // We've exhausted `b`, just push the def from `a` and move on to the next.
+                    push(a_def, &mut a_constraints_iter, &mut merged);
+                    opt_a_def = a_defs_iter.next();
+                }
+                (None, Some(b_def)) => {
+                    // We've exhausted `a`, just push the def from `b` and move on to the next.
+                    push(b_def, &mut b_constraints_iter, &mut merged);
+                    opt_b_def = b_defs_iter.next();
+                }
+                (None, None) => break,
+            }
+        }
+        merged
+    }
+
+    /// Get iterator over visible definitions with constraints.
+    pub(super) fn iter_visible_definitions(&self) -> DefinitionIdWithConstraintsIterator {
```

Comment:
```suggestion
    pub(super) fn visible_definitions(&self) -> DefinitionIdWithConstraintsIterator {
```

The `iter` prefix is less common when iterating named parts of a struct as we do here. Just `visible_definitions` seems enough. The return type already specifies that it is an iterator.